### PR TITLE
Attributes branch for review

### PR DESCRIPTION
**Pull request by progrium**

### CODE REVIEW

File: discover/backend_etcd.go [hunk 1]
```diff
@@ -69,27 +92,35 @@ func (b *EtcdBackend) responseToUpdate(resp *store.Response) *ServiceUpdate {
...
 
-func (b *EtcdBackend) Register(name string, addr string, attrs map[string]string) error {
-	_, err := b.Client.Set(fmt.Sprintf("/services/%s/%s", name, addr), addr, HeartbeatIntervalSecs+MissedHearbeatTTL)
-	return err
+func (b *EtcdBackend) Register(name, addr string, attrs map[string]string) error {
+	attrsJson, err1 := json.Marshal(attrs)
+	if err1 != nil {
+		return err1
+	}
+	_, err2 := b.Client.Set(servicePath(name, addr), string(attrsJson), HeartbeatIntervalSecs+MissedHearbeatTTL)
```
Comment (titanous) [inline discover/backend_etcd.go]: Call both of these errors `err` (this isn't Erlang :P)

File: discover/backend_etcd.go [hunk 2]
```diff
@@ -9,10 +10,22 @@ import (
 	"github.com/coreos/go-etcd/etcd"
 )
 
+const (
+	KeyPrefix = "/discover"
+)
```
Comment (titanous) [inline discover/backend_etcd.go]:
``` go
const KeyPrefix = "/discover"
```

File: discover/backend_etcd.go [hunk 3]
```diff
@@ -9,10 +10,22 @@ import (
...
+
 type EtcdBackend struct {
 	Client *etcd.Client
 }
 
+func servicePath(name, addr string) string {
+	if addr != "" {
+		return fmt.Sprintf("%s/services/%s/%s", KeyPrefix, name, addr)
+	} else {
+		return fmt.Sprintf("%s/services/%s", KeyPrefix, name)
```
Comment (titanous) [inline discover/backend_etcd.go]: You use simple concatenation [below](https://github.com/flynn/go-discover/pull/5/files#diff-8e82ad9b65ab1317f873c67f8fafaa10R29) and `fmt` here. For consistency, pick one or the other. I don't care either way, but `fmt.Sprintf` is almost certainly slower.

File: discover/backend_etcd_test.go [hunk 4]
```diff
@@ -9,52 +9,70 @@ import (
...
-}
-
 func deleteService(client *etcd.Client, service string, addr string) {
 	client.Delete(fmt.Sprintf("/services/%s/%s", service, addr))
 }
 
-func TestEtcdBackend_RegisterAndUnregister(t *testing.T) {
+const (
+	NoAttrService = "null"
+)
```
Comment (titanous) [inline discover/backend_etcd_test.go]: Change this to a one-line const.

File: discover/backend_etcd.go [hunk 5]
```diff
@@ -46,22 +60,31 @@ func (s *etcdStream) Chan() chan *ServiceUpdate { return s.ch }
...
-	serviceName := respList[2]
-	if ("SET" == resp.Action && resp.NewKey) || "GET" == resp.Action {
+	serviceName := splitKey[3]
+	serviceAddr := splitKey[4]
+	if "GET" == resp.Action || ("SET" == resp.Action && (resp.NewKey || resp.Value != resp.PrevValue)) {
+		// GET is because getCurrentState returns responses of Action GET.
+		// some SETs are heartbeats, so we ignore SETs where value didn't change.
+		var serviceAttrs map[string]string
+		err := json.Unmarshal([]byte(resp.Value), &serviceAttrs)
+		if err != nil {
```
Comment (titanous) [inline discover/backend_etcd.go:71]: You could collapse this if you wanted to (I use both styles):

``` go
if err := json.Unmarshal([]byte(resp.Value), &serviceAttrs); err != nil { 
```

File: discover/backend_etcd.go (hunk 3 above)
Comment (titanous) [inline discover/backend_etcd.go]: The idiomatic style would be to invert the statement and then outdent the else block:

``` go
if addr == "" {
    return fmt.Sprintf("%s/services/%s", KeyPrefix, name)
}
return fmt.Sprintf("%s/services/%s/%s", KeyPrefix, name, addr)
```

File: discover/backend_etcd.go [hunk 6]
```diff
@@ -26,6 +39,7 @@ func (b *EtcdBackend) Subscribe(name string) (UpdateStream, error) {
 				stream.ch <- update
 			}
 		}
+		stream.ch <- &ServiceUpdate{}
```
Comment (titanous) [inline discover/backend_etcd.go:38]: Could also send `nil` as the sentinel.

File: discover/client.go [hunk 7]
```diff
@@ -29,10 +29,21 @@ type ServiceSet struct {
...
+func (s *ServiceSet) bind(updates chan *ServiceUpdate) chan bool {
+	// current is an event when enough service updates have been
+	// received to bring us to "current" state (when subscribed)
+	current := make(chan bool)
 	go func() {
+		isCurrent := false
 		for update := range updates {
-			// TODO: apply filters
+			if update.Addr == "" && update.Name == "" && !isCurrent {
+				current <- true
```
Comment (titanous) [inline discover/client.go]: I'd use a `chan struct{}` and then just `close(current)`.

File: discover/client.go [hunk 8]
```diff
@@ -97,11 +120,35 @@ func (s *ServiceSet) OfflineAddrs() []string {
 	return list
 }
 
+func (s *ServiceSet) Select(attrs map[string]string) []*Service {
+	s.serMutex.Lock()
+	defer s.serMutex.Unlock()
+	list := make([]*Service, 0, len(s.services))
+outter:
```
Comment (titanous) [inline discover/client.go]: `s/outter/outer/`

File: discover/backend_etcd.go (hunk 5 above)
Comment (progrium) [inline discover/backend_etcd.go:71]: since it's rather long i'll leave it, but i need to get in the habit of doing that more.

Comment (titanous) [inline discover/backend_etcd.go:71]: Yeah, just a reminder.

File: discover/backend_etcd.go (hunk 6 above)
Comment (progrium) [inline discover/backend_etcd.go:38]: Wasn't sure if rpcplus streams support nil, but when I think about it I don't see why not.

Comment (progrium) [inline discover/backend_etcd.go:38]: actually it's gob that doesn't like it: `gob: cannot encode nil pointer of type *discover.ServiceUpdate`

**Comment by titanous** [comment]
LGTM.

**Pull request merged**